### DISCUSSION TITLE
Unable to cast ItemStack to ItemStackDataSaverInterface

### DISCUSSION THREAD

**Discussion by TechPro424** [FabricMC/fabric-api#3035] [Mod Dev Support] [answered]
I'm trying to store some data in the NBT of an `ItemStack` using a command, a mixin and an interface by following these 2 tutorials by Kaupenjoe and adapting them for 1.19.4:
https://youtu.be/U7j4bl_UAII
https://www.youtube.com/watch?v=jak1yHxjKq0&t=336s

However, in my command, when I try to cast an `ItemStack` to my `ItemStackDataSaverInterface`, I get the error `Cannot cast from ItemStack to ItemStackDataSaverInterface`

Why is that? Please help me out
All relevant code is in the following Gist: https://gist.github.com/TechPro424/2d4bd6ac27dc421d44e34bfa7b394761

**Comment by warjort** [comment, marked as answer]
Maybe you meant to do this?
https://fabricmc.net/wiki/tutorial:interface_injection
Otherwise, the ItemStack doesn't implement your interface at compile time only at runtime.

Even without that, you can fool the compiler by first casting to a java.lang.Object:
````
ItemStackDataSaverInterface stackAsInterface = (ItemStackDataSaverInterface) (Object) stack;
````

**Reply by TechPro424** [reply]
But shouldn't you be able to cast any object to any interface in Java?

**Reply by apple502j** [reply]
No; it probably compiles, but will fail when executed.

**Reply by TechPro424** [reply]
Ok, I've casted it to my interface after casting it to an object, the way @warjort has mentioned, but now, each time I try to use `ItemStack.getNbt().asString()` in the command, it throws an error saying `An unexpected error occurred trying to execute that command`
But `ItemStack.getNbt()` works perfectly fine

**Reply by warjort** [reply]
You need to show the full error from the log.

I've warned you before, that if you don't post all the relevant information, the most likely outcome is your question just gets ignored.
The easiest way to do that is to put the code on github (as a project not a gist).

I'm guessing (I shouldn't have to guess) you forgot to define your mixin in your mixin configuration? Since you haven't shown that in anything you posted so far.
That would mean your cast is failing at runtime because the ItemStack does not in fact implement the interface.

**Reply by TechPro424** [reply]
> You need to show the full error from the log.
> 
> I've warned you before, that if you don't post all the relevant information, the most likely outcome is your question just gets ignored. The easiest way to do that is to put the code on github (as a project not a gist).
> 
> I'm guessing (I shouldn't have to guess) you forgot to define your mixin in your mixin configuration? Since you haven't shown that in anything you posted so far. That would mean your cast is failing at runtime because the ItemStack does not in fact implement the interface.

If you mean `latest.log`, it's logged the same as the error in game: `[13:28:28] [Render thread/INFO] (Minecraft) [System] [CHAT] An unexpected error occurred trying to execute that command`

And I have defined it, I double-checked

**Reply by warjort** [reply]
Minecraft commands don't log their real errors unless you add code to tell them to do it.
Add a try/catch around your code and log the error.

**Reply by warjort** [reply]
Or use a debugger to look at the error.

**Reply by TechPro424** [reply]
> Minecraft commands don't log their real errors unless you add code to tell them to do it. Add a try/catch around your code and log the error.

The error is an NPE, the item NBT is null
And logging the NBT using that method seems to be useless anyway, because it doesn't seem to get the values of my tag (I checked with an enchanted sword, to ensure that the NBT was not null)

Sorry for wasting your time, my bad :(
